### PR TITLE
Validation and error improvements

### PR DESCRIPTION
**Pull request by sambauers**
Refer to #3 for background.

### CODE REVIEW

**Comment by sambauers** [comment]
This bumps the version to 0.3.2, so will require a release if merged.

**Comment by sambauers** [comment]
@trev, thanks for the feedback - here's what I have done:
- Silent failure on `to_s` is gone, replaced with exception raising as appropriate
- No more optional validation on `to_s`, it now raises meaningful exceptions in certain cases though
- No longer restricting transaction code to a list of numbers, just accepting two numbers now
- Pry dependency taken out

Other improvements and changes in a nutshell, besides the validation additions:
- `Aba.add_transaction` now accepts a `Hash` of attributes to be passed to a new `Aba::Transaction` object
- `Aba.new` also accepts an optional second parameter which is an array of transactions to add
- `Aba.all_valid?` returns boolean flagging if there any validation issue with the object or a contained transaction
- `Aba.all_errors` returns a structured `Hash` containing ABA object and transaction validation errors
- Access to the transactions is now easier due to transactions now being stored in an indexed `Hash`, so things like `aba.transactions[3].bsb = '123-456'` are now possible
- Added some type conversion so that setting attributes as integers doesn't break
- Strip dashes from account_number and trace_account_number on get (specification says to strip dashes if string is longer than 9 chars - says nothing about spaces though, which are allowed)

**Comment by sambauers** [comment]
I think I have done everything I set out to do now. Waiting for feedback on what parts need improvement, or excising.

**Comment by trev** [comment]
Looks great @sambauers, I'll have a look tonight as will @andrba I'm sure. :+1:

**Comment by andrba** [comment]
If we look at the instance methods of the Aba class we'll see among the others something like this:

``` ruby
Aba.instance_methods.grep(/valid|errors/)
=> [:transactions_valid?, :all_valid?, :all_errors, :errors, :errors=, :valid?]
```

Is there any use of `valid?` and `errors` at all? Maybe it's worth overriding them? Say `all_valid?` will be simply `valid?` and `errors` will return all errors (without validation)? Having so many error and validation methods seems to be confusing.

File: lib/aba.rb [hunk 1]
```diff
@@ -5,40 +5,95 @@
...
+    unless transactions.nil? || transactions.empty?
+      transactions.to_a.each do |t|
+        self.add_transaction(t) unless t.nil? || t.empty?
+      end
+    end
 
     yield self if block_given?
   end
 
   def to_s
```
Comment (andrba) [inline lib/aba.rb:52]: Everything dangerous must end with a `!` sign. Should be named `to_s!`

Comment (trev) [inline lib/aba.rb:52]: I'm not sure about that @andrba. The `#to_s` method is not destructive in any way. It doesn't modify any object, just returns a string representation.

**Comment by trev** [comment]
I'll have to agree with @andrba on the validation methods. Otherwise I'd say it's looking pretty damn good.

File: lib/aba.rb (hunk 1 above)
Comment (andrba) [inline lib/aba.rb:52]: Probably we can leave it as it is. I'm used to the rails style where bangs can indicate that a method can throw an exception.

**Comment by trev** [comment]
Hey @sambauers, if you want to reduce the validation methods to just `#valid?` and `#errors` I'll merge.

**Comment by sambauers** [comment]
Won't that involve stomping on the existing `valid?` method and `errors` object?

They are included via `include Aba::Validations` in `aba/lib/aba.rb`

I would have to rename those methods in the `Validations` module - is that OK?

**Comment by sambauers** [comment]
And I think the `to_s` method should remain named as it is. It is not destructive at all.

**Comment by trev** [comment]
I think this looks pretty good! I think the errors could probably be refined a bit more but I'll submit my own PR later. Thanks for your contributions @sambauers. :+1:

**Pull request merged**